File: debop4s-timeperiod/src/test/scala/com/github/debop4s/timeperiod/tests/timeranges/HalfyearRangeCollectionTest.scala
package com.github.debop4s.timeperiod.tests.timeranges

import com.github.debop4s.timeperiod._
import com.github.debop4s.timeperiod.tests.AbstractTimePeriodTest
import com.github.debop4s.timeperiod.timerange.{HalfyearRange, HalfyearRangeCollection}
import com.github.debop4s.timeperiod.utils.Times
import com.github.debop4s.timeperiod.utils.Times._

/**
 * HalfyearRangeCollectionTest
 * Created by debop on 2014. 2. 15.
 */
class HalfyearRangeCollectionTest extends AbstractTimePeriodTest {

    test("yearBaseMonth") {
        val moment = Times.asDate(2009, 2, 15)
        val year = yearOf(moment.getYear, moment.getMonthOfYear)
        val halfyears = HalfyearRangeCollection(moment, 3)

        halfyears.start should equal(asDate(year, 1, 1))
    }

    test("single Halfyear") {
        val startYear = 2004
        val startHalfyear = Halfyear.Second

        val halfyears = HalfyearRangeCollection(startYear, startHalfyear, 1)

        halfyears.halfyearCount should equal(1)
        halfyears.startHalfyear should equal(startHalfyear)

        halfyears.endYear should equal(startYear)
        halfyears.endHalfyear should equal(startHalfyear)

        val halfyearList = halfyears.halfyears
        halfyearList.size should equal(1)
        halfyearList.head.isSamePeriod(new HalfyearRange(startYear, startHalfyear)) should equal(true)

        val halfyearView = halfyears.halfyearsView
        halfyearView.size should equal(1)
        halfyearView.head.isSamePeriod(new HalfyearRange(startYear, startHalfyear)) should equal(true)
    }

    test("first calendar halfyears") {
        val startYear = 2004
        val startHalfyear = Halfyear.First
        val halfyearCount = 3

        val halfyears = HalfyearRangeCollection(startYear, startHalfyear, halfyearCount)

        halfyears.halfyearCount should equal(halfyearCount)
        halfyears.startHalfyear should equal(startHalfyear)
        halfyears.endYear should equal(startYear + 1)
        halfyears.endHalfyear should equal(Halfyear.First)

        val halfyearList = halfyears.halfyears

        halfyearList.size should equal(halfyearCount)
        halfyearList(0).isSamePeriod(HalfyearRange(startYear, Halfyear.First)) should equal(true)
        halfyearList(1).isSamePeriod(HalfyearRange(startYear, Halfyear.Second)) should equal(true)
        halfyearList(2).isSamePeriod(HalfyearRange(startYear + 1, Halfyear.First)) should equal(true)

        val halfyearView = halfyears.halfyearsView

        halfyearView.size should equal(halfyearCount)
        halfyearView(0).isSamePeriod(HalfyearRange(startYear, Halfyear.First)) should equal(true)
        halfyearView(1).isSamePeriod(HalfyearRange(startYear, Halfyear.Second)) should equal(true)
        halfyearView(2).isSamePeriod(HalfyearRange(startYear + 1, Halfyear.First)) should equal(true)
    }

    test("secondCalendarHalfyears") {
        val startYear = 2004
        val startHalfyear = Halfyear.Second
        val halfyearCount = 3

        val halfyears = HalfyearRangeCollection(startYear, startHalfyear, halfyearCount)

        halfyears.halfyearCount should equal(halfyearCount)
        halfyears.startHalfyear should equal(startHalfyear)
        halfyears.endYear should equal(startYear + 1)
        halfyears.endHalfyear should equal(Halfyear.Second)

        val halfyearList = halfyears.halfyears

        halfyearList.size should equal(halfyearCount)
        halfyearList(0).isSamePeriod(HalfyearRange(startYear, Halfyear.Second)) should equal(true)
        halfyearList(1).isSamePeriod(HalfyearRange(startYear + 1, Halfyear.First)) should equal(true)
        halfyearList(2).isSamePeriod(HalfyearRange(startYear + 1, Halfyear.Second)) should equal(true)
    }
}

File: debop4s-timeperiod/src/test/java/debop4s/timeperiod/tests/timeranges/HalfyearRangeCollectionTest.java
package debop4s.timeperiod.tests.timeranges;

import debop4s.timeperiod.Halfyear;
import debop4s.timeperiod.tests.TimePeriodTestBase;
import debop4s.timeperiod.timerange.HalfyearRange;
import debop4s.timeperiod.timerange.HalfyearRangeCollection;
import debop4s.timeperiod.utils.Times;
import lombok.extern.slf4j.Slf4j;
import org.joda.time.DateTime;
import org.junit.Test;

import java.util.List;

import static org.fest.assertions.Assertions.assertThat;

/**
 * debop4s.timeperiod.test.timeranges.HalfyearRangeCollectionFunSuite
 *
 * @author 배성혁 sunghyouk.bae@gmail.com
 * @since 13. 5. 25. 오후 4:42
 */
@Slf4j
public class HalfyearRangeCollectionTest extends TimePeriodTestBase {

    @Test
    public void yearBaseMonthTest() {

        DateTime moment = Times.asDate(2009, 2, 15);
        int year = Times.yearOf(moment.getYear(), moment.getMonthOfYear());
        HalfyearRangeCollection halfyears = new HalfyearRangeCollection(moment, 3);

        assertThat(halfyears.getStart()).isEqualTo(Times.asDate(year, 1, 1));
    }

    @Test
    public void singleHalfyearTest() {
        final int startYear = 2004;
        final Halfyear startHalfyear = Halfyear.Second;

        HalfyearRangeCollection halfyears = new HalfyearRangeCollection(startYear, startHalfyear, 1);

        assertThat(halfyears.getHalfyearCount()).isEqualTo(1);
        assertThat(halfyears.getStartHalfyear()).isEqualTo(startHalfyear);
        assertThat(halfyears.getEndYear()).isEqualTo(startYear);
        assertThat(halfyears.getEndHalfyear()).isEqualTo(startHalfyear);

        List<HalfyearRange> halfyearList = halfyears.halfyears();
        assertThat(halfyearList.size()).isEqualTo(1);
        assertThat(halfyearList.get(0).isSamePeriod(new HalfyearRange(2004, Halfyear.Second))).isTrue();
    }

    @Test
    public void firstCalendarHalfyears() {
        final int startYear = 2004;
        final Halfyear startHalfyear = Halfyear.First;
        final int halfyearCount = 3;

        HalfyearRangeCollection halfyears = new HalfyearRangeCollection(startYear, startHalfyear, halfyearCount);

        assertThat(halfyears.getHalfyearCount()).isEqualTo(halfyearCount);
        assertThat(halfyears.getStartHalfyear()).isEqualTo(startHalfyear);
        assertThat(halfyears.getEndYear()).isEqualTo(startYear + 1);
        assertThat(halfyears.getEndHalfyear()).isEqualTo(Halfyear.First);

        List<HalfyearRange> halfyearList = halfyears.halfyears();

        assertThat(halfyearList.size()).isEqualTo(halfyearCount);
        assertThat(halfyearList.get(0).isSamePeriod(new HalfyearRange(2004, Halfyear.First))).isTrue();
        assertThat(halfyearList.get(1).isSamePeriod(new HalfyearRange(2004, Halfyear.Second))).isTrue();
        assertThat(halfyearList.get(2).isSamePeriod(new HalfyearRange(2005, Halfyear.First))).isTrue();
    }

    @Test
    public void secondCalendarHalfyears() {
        final int startYear = 2004;
        final Halfyear startHalfyear = Halfyear.Second;
        final int halfyearCount = 3;

        HalfyearRangeCollection halfyears = new HalfyearRangeCollection(startYear, startHalfyear, halfyearCount);

        assertThat(halfyears.getHalfyearCount()).isEqualTo(halfyearCount);
        assertThat(halfyears.getStartHalfyear()).isEqualTo(startHalfyear);
        assertThat(halfyears.getEndYear()).isEqualTo(startYear + 1);
        assertThat(halfyears.getEndHalfyear()).isEqualTo(Halfyear.Second);

        List<HalfyearRange> halfyearList = halfyears.halfyears();

        assertThat(halfyearList.size()).isEqualTo(halfyearCount);
        assertThat(halfyearList.get(0).isSamePeriod(new HalfyearRange(2004, Halfyear.Second))).isTrue();
        assertThat(halfyearList.get(1).isSamePeriod(new HalfyearRange(2005, Halfyear.First))).isTrue();
        assertThat(halfyearList.get(2).isSamePeriod(new HalfyearRange(2005, Halfyear.Second))).isTrue();
    }

}
